bundler/lib/bundler/resolver.rb

```diff
@@ -37,9 +37,17 @@ def setup_solver
       root_version = Resolver::Candidate.new(0)
 
       @all_specs = Hash.new do |specs, name|
-        specs[name] = source_for(name).specs.search(name).reject do |s|
-          s.dependencies.any? {|d| d.name == name && !d.requirement.satisfied_by?(s.version) } # ignore versions that depend on themselves incorrectly
-        end.sort_by {|s| [s.version, s.platform.to_s] }
+        source = source_for(name)
+        matches = source.specs.search(name)
+
+        # Don't bother to check for circular deps when no dependency API are
+        # available, since it's too slow to be usable. That edge case won't work
+        # but resolution other than that should work fine and reasonably fast.
+        if source.respond_to?(:dependency_api_available?) && source.dependency_api_available?
```

Comment:
Ah ok. Nice. I couldn't figure out where this check belonged. I tried putting something on remote_specification that would tell it not to use the source to load dependencies, but it was a bit ugly. I'll see if I can write a test for this and we can merge.

Reply:
I'm leaving for a few weeks, so feel free to proceed as you see fit. We could close this PR, cherry-pick these two commits there, and get the issue fix there?